spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameterManager.scala

```diff
@@ -0,0 +1,333 @@
+    BlockManagerWrapper.putSingle(getGradientExecutorId(),
+      _gradientsExecutor, StorageLevel.MEMORY_AND_DISK, tellMaster = false)
+
+    val blockId = getWeightBlockId(executorId)
+    val fp16param = new FP16CompressedTensor[T](length)(_classTag)
+    fp16param.compress(0, parameter, start, length)
+    BlockManagerWrapper.putBytes(blockId, fp16param.bytes(), StorageLevel.MEMORY_ONLY_SER)
+  }
+
+  /** Aggregate gradients hosted in one executor */
+  def aggregateLocalGradient[T: ClassTag]() : Tensor[T] = {
+    val blockIds = master.getBlockId(executorId)
+    val gradientBuffer = new Array[Tensor[T]](blockIds.size)
+    Engine.default.invokeAndWait((0 until blockIds.size).map(tid => () => {
+      gradientBuffer(tid) = getLocalParameter(blockIds(tid))
+    }))
+
+    val poolSize = Engine.default.getPoolSize
+    val innerTaskSize = size / poolSize
+    val innerExtraSize = size % poolSize
+    val availableTask = if (innerTaskSize == 0) innerExtraSize else poolSize
+
+    Engine.default.invokeAndWait((0 until availableTask).map(tid => () => {
+      val innerStart = tid * innerTaskSize + math.min(tid, innerExtraSize)
+      val innerLength = innerTaskSize + (if (tid < innerExtraSize) 1 else 0)
+      var i = 1
+      while (i < gradientBuffer.length) {
+        gradientBuffer(0).narrow(1, innerStart + 1, innerLength)
+          .add(gradientBuffer(i).narrow(1, innerStart + 1, innerLength))
+        i += 1
+      }
+    }))
+    master.clearBlockId(executorId)
+    gradientBuffer(0)
+  }
+
+  /**
+    * Slice aggregated gradients into chunks, and mark each chunk to be sent
+    * to the appropriate parameter node, and put it in the block manager.
+    *
+    * @param parameter A Tensor that contains gradients computed on the entire model on a single
+    *                  node.
+    */
+  def putGradients[T: ClassTag](parameter: Tensor[T]): Unit = {
+    val _classTag = classTag[T]
+    var pid = 0
+    while (pid < executorNum) {
+      val start = pid * taskSize + math.min(pid, extraSize)
+      val length = taskSize + (if (pid < extraSize) 1 else 0)
+      val blockId = getGradientBlockId(executorId, pid)
+      val block = BlockManagerWrapper.getLocalBytes(blockId)
+      if (block.isDefined) {
+        val fp16param = new FP16CompressedTensor[T](block.get)(_classTag)
+        fp16param.compress(0, parameter, start, length)
+      } else {
+        val fp16param = new FP16CompressedTensor[T](length)(_classTag)
+        fp16param.compress(0, parameter, start, length)
+        BlockManagerWrapper.putBytes(blockId, fp16param.bytes(), StorageLevel.MEMORY_ONLY_SER)
+      }
+
+      pid += 1
+    }
+  }
+
+  /**
+    * Retrieve gradients for the slice of the model that this node is responsible for from all the
+    * other nodes. A new thread is created for each separate node. The gradients are then summed
+    * and then stored in decompressed form in blockmanager.
+    */
+  def aggregrateGradientParition[T: ClassTag](): Unit = {
```

Comment:
what's the different between aggregateLocalGradient and aggregrateGradientParition?